src/expr/src/linear.rs

```diff
@@ -533,4 +546,51 @@ impl MapFilterProject {
         // This should probably resemble existing scalar cse.
         unimplemented!()
     }
+
+    /// Removes unused expressions from `self.expressions`.
+    ///
+    /// Expressions are "used" if they are relied upon by any output columns
+    /// or any predicates, even transitively. Any expressions that are not
+    /// relied upon in this way can be discarded.
+    pub fn remove_undemanded(&mut self) {
+        // Determine the demanded expressions to remove irrelevant ones.
+        let mut demand = std::collections::HashSet::new();
+        for (_index, pred) in self.predicates.iter() {
+            demand.extend(pred.support());
+        }
+        demand.extend(self.projection.iter().cloned());
+        for index in (0..self.expressions.len()).rev() {
+            if demand.contains(&(self.input_arity + index)) {
+                demand.extend(self.expressions[index].support());
+            }
+        }
+
+        // Expressions that are not demanded can be removed, and all
+        // remaining indexes updated.
+        let mut remap = HashMap::new();
+        for index in 0..self.input_arity {
+            remap.insert(index, index);
+        }
+        // Retain demanded expressions, and record their new locations.
+        let mut index = self.input_arity;
+        self.expressions.retain(|_expr| {
+            let demanded = demand.contains(&index);
+            if demanded {
+                remap.insert(index, remap.len());
+            }
+            index += 1;
+            demanded
+        });
+        // Update column identifiers.
+        for expr in self.expressions.iter_mut() {
+            expr.permute_map(&remap);
+        }
+        for (_pos, pred) in self.predicates.iter_mut() {
+            // NB: leaving `_pos` for now.
```

Comment:
I'm not sure if you intended to change this before merging but I'm having trouble figuring out if this is ok to do because when we evaluate it seems like we need position to be less than or equal to`input_arity + expressions.len()`

based on:
```
110         for (support, predicate) in self.predicates.iter() {
111             while self.input_arity + expression < *support {
112                 datums.push(self.expressions[expression].eval(&datums[..], &arena)?);
113                 expression += 1;
114             }
115             if predicate.eval(&datums[..], &arena)? != Datum::True {
116                 return Ok(None);
117             }
118         }
```

again, might be missing something

Reply:
No, you seem totally right. I was just imagining it would be a late applied predicate. I'll fix that up now.

Reply:
cool. yeah I'm trying to go piecemeal here just trying to understand invariants in the code as a entry point to understanding whats going on